controller/src/main/java/io/pravega/controller/store/stream/tables/SegmentRecord.java

```diff
@@ -9,62 +9,87 @@
      * @param segmentTable segment table
      * @param number segment number to read
      * @return returns segment record
      */
-    static Optional<SegmentRecord> readRecord(final byte[] segmentTable, final int number) {
-        int offset = number * SegmentRecord.SEGMENT_RECORD_SIZE;
+    static Optional<SegmentRecord> readRecord(final byte[] segmentIndex, final byte[] segmentTable, final int number) {
+        Optional<SegmentIndexRecord> indexRecord = SegmentIndexRecord.readRecord(segmentIndex, number);
+        return indexRecord.map(segmentIndexRecord -> {
+            if (indexRecord.get().getSegmentOffset() >= segmentTable.length) {
+                // index may be ahead of segment table. So we may not be able to read the record.
+                return null;
+            } else {
+                return parse(segmentTable, segmentIndexRecord.getSegmentOffset());
+            }
+        });
+    }
+
+    /**
+     * Method to read latest record in segment table.
+     * Note: index may be ahead of segment table. So we will find the last record in segment table and ignore index entries
+     * that are not written to segment table yet.
+     * @param segmentIndex segment index
+     * @param segmentTable segment table
+     * @return returns latest segment record in segment table
+     */
+    static Optional<SegmentRecord> readLatest(final byte[] segmentIndex, final byte[] segmentTable) {
+        Optional<SegmentIndexRecord> indexRecord = SegmentIndexRecord.readLatestRecord(segmentIndex);
 
-        if (offset >= segmentTable.length) {
-            return Optional.empty();
+        while (indexRecord.isPresent()) {
+            int segmentNumber = indexRecord.get().getSegmentNumber();
+            if (indexRecord.get().getSegmentOffset() >= segmentTable.length) {
+                indexRecord = SegmentIndexRecord.readRecord(segmentIndex, segmentNumber - 1);
+            } else {
+                return Optional.of(parse(segmentTable, indexRecord.get().getSegmentOffset()));
+            }
         }
-        return Optional.of(parse(segmentTable, offset));
+
+        return Optional.empty();
     }
 
+    @SneakyThrows
     private static SegmentRecord parse(final byte[] table, final int offset) {
-        return new SegmentRecord(BitConverter.readInt(table, offset),
-                BitConverter.readInt(table, offset + Integer.BYTES),
-                BitConverter.readLong(table, offset + 2 * Integer.BYTES),
-                toDouble(table, offset + 2 * Integer.BYTES + Long.BYTES),
-                toDouble(table, offset + 2 * Integer.BYTES + Long.BYTES + Double.BYTES));
+        InputStream bas = new ByteArrayInputStream(table, offset, table.length - offset);
+        return SERIALIZER.deserialize(bas);
     }
 
-    private static double toDouble(byte[] b, int offset) {
-        return Double.longBitsToDouble(BitConverter.readLong(b, offset));
+    @SneakyThrows
+    byte[] toByteArray() {
```

Comment:
In order to avoid all this byte array copying everywhere, I recommend changing this to return `ArrayView` and keep using it until you really have no choice but copy it to a byte array. For instance, many APIs allow taking a byte array with offset + length, so no need for any additional copies.

Reply:
done